### PR TITLE
Add listener to DOMContentLoaded when document readyState is loading

### PR DESCRIPTION
**Pull request by safaiyeh**
## Problem
The document load event waits for the browser to load all resources before executing [(reference)](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event). The Magic Eden web app is very resource heavily with all our images, videos, and gifs which caused the wallet connection to be delayed. The quick patch we used was to listen to the [DOMContentLoaded event](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event) so wallets can be connected before all the resources loaded.

## Solution 
@ArpitAgrawal321 suggested in https://github.com/solana-labs/wallet-adapter/issues/317 to fully use DOMContentLoaded for this case. @steveluscher [suggested](https://github.com/solana-labs/wallet-adapter/issues/317#issuecomment-1050538881) to also handle the in-between states of the document loading and finished loading.

This PR adds a `DOMContentLoaded` listener when the document state is loading while still watching the window `load` event in the case that the document is in the `complete` state

### CODE REVIEW

**Comment by safaiyeh** [comment]
> > while still watching the window `load` event in the case that the document is in the `complete` state
> 
> Looks good, but I'm wondering about this part. It looks like it early returns if the state is `complete`?

Ah right, misspoke there. If state is complete the load event wouldn’t happen again, so it wouldn’t need the event listener.

**Comment by jordansexton** [comment]
Makes sense, thanks! I'll patch release in the morning.

**Review by steveluscher — CHANGES_REQUESTED** [review]
Thanks for drumming this up! Here's how I understand this implementation:

* IF the document is `ready`, detect.
  * IF detection failed, poll indefinitely until wallet is found
* ELSE IF document is loading, listen for document to become `ready`
  * WHEN `ready`, detect
  * IF detection failed, poll indefinitely until wallet is found
* ELSE listen for `load` event
  * WHEN `load` event is detected, detect
  * IF detection failed, poll indefinitely until wallet is found

The interesting thing to not about this is that the decision of when to start polling depends largely on when this code is run.

```
// DOMContentLoaded has not fired by the time you call this code?

                       polling starts here
                       V
loading --------> interactive --------> complete --------> load fires
            (aka DOMContentLoaded)

// DOMContentLoaded already fired by the time you call this code?

                                                           polling starts here
                                                           V
loading --------> interactive --------> complete --------> load fires
            (aka DOMContentLoaded)
```

I think the fix here is to introduce a check for `readyState === 'interactive'`. The `DOMContentLoaded` event and the `interactive` ready state are synonymous. Here, check it out: https://jsbin.com/yoniharigo/1/edit?html,js,console,output

tl;dr:

* `'loading'` implies that neither `DOMContentLoaded` nor the `load` event have fired.
* `'interactive'` implies that `DOMContentLoaded` has fired but the `load` event has not.
* `'complete'` implies that `DOMContentLoaded` and the `load` event have fired.

Also, if we start polling at `DOMContentLoaded`, this code does _not_ set up the `load` event listener. This means that we have, at most, 1000ms precision on detecting the wallet. If there's a wallet that typically becomes available on `load`, we could be late detecting it by up to 1000ms, which is huge in terms of the startup of a website.

Alright, that was a lot of talk. What's my feedback? Implement the missing functionality:

* [X] If `readyState` is loading, set up the `DOMContentLoaded` listener…
  * [x] …_and_ set up the `load` event listener. 
* [ ] If `readyState` is `interactive`, fire the `poll()` method.
  * [X] If detection failed, set up the interval…
  * [ ] …_and_ set up the `load` event listener.
* [X] If `readyState` is `complete`, fire the `poll()` method.
  * [X] If detection failed, set up the interval.
* [ ] If at any point detection succeeds, remove all of the listeners and cancel all of the intervals.

I'll probably add a bunch of stuff to this later, like:

* Exponential backoff of the poll interval (eg. poll less and less often as we get further from startup).
* The use of `requestIdleCallback` to make sure polling never runs when hi-pri startup work is underway.
* Cooperative multitasking, to make sure that we yield the thread between each wallet detector run.

**Comment by steveluscher** [comment]
Haha. Missed the merge by 4 minutes.

**Comment by steveluscher** [comment]
Ignore everything above, I'll send a follow up.

**Comment by steveluscher** [comment]
Closes #317.
